### PR TITLE
Discovery cell project category experiment

### PR DESCRIPTION
**Pull request by ifbarrera**
# What

Adding project category and `Projects We Love` tag to the discovery postcard cells, as an A/B experiment.

simulator screen shot - iphone 8 - 2018-05-07 at 17 03 52

simulator screen shot - iphone 5s - 2018-05-07 at 17 03 18

### CODE REVIEW

File: Kickstarter-iOS/ViewModels/DiscoveryProjectCategoryViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,53 @@
+//
+//  DiscoveryProjectCategoryViewModel.swift
+//  Kickstarter-iOS
+//
+//  Created by Isabel Barrera on 4/30/18.
+//  Copyright © 2018 Kickstarter. All rights reserved.
+//
```
Comment (Scollaco) [inline Kickstarter-iOS/ViewModels/DiscoveryProjectCategoryViewModel.swift]: Could you delete this header comments?

File: Kickstarter-iOS/Views/DiscoveryProjectCategoryView.swift [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  @IBOutlet private weak var blurView: UIImageView!
+  @IBOutlet private weak var categoryStackView: UIStackView!
+  @IBOutlet private weak var categoryViewImageView: UIImageView!
+  @IBOutlet private weak var categoryViewLabel: UILabel!
+
+  public required init?(coder aDecoder: NSCoder) {
+    super.init(coder: aDecoder)
+  }
+
+  internal func configureWith(name: String, imageNameString: String) {
```
Comment (Scollaco) [inline Kickstarter-iOS/Views/DiscoveryProjectCategoryView.swift]: Here we could use only one input and pass the name and imageName, like: `viewModel.inputs.configureWith(name: name, imageNameString: imageNameString)`.

File: Library/Config+Helpers.swift [hunk 3]
```diff
@@ -0,0 +1,22 @@
+//
+//  Config+Helpers.swift
+//  Kickstarter-iOS
+//
+//  Created by Isabel Barrera on 5/7/18.
+//  Copyright © 2018 Kickstarter. All rights reserved.
+//
```
Comment (Scollaco) [inline Library/Config+Helpers.swift]: Could you delete these comments as well?

File: Library/Config+Helpers.swift [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+import Foundation
+import KsApi
+
+extension Experiment.Name {
```
Comment (Scollaco) [inline Library/Config+Helpers.swift]: Nice extension! 👍

File: KsApi/models/Project.swift [hunk 5]
```diff
@@ -6,6 +6,7 @@ import Prelude
 public struct Project {
 
   public private(set) var blurb: String
+  public private(set) var staffPick: Bool
```
Comment (Scollaco) [inline KsApi/models/Project.swift]: A smaller thing here... We try to keep the properties of the struct alphabetized (just like we do with the list of inputs/outputs functions).

File: Library/ViewModels/DiscoveryPostcardViewModel.swift [hunk 6]
```diff
@@ -346,6 +403,11 @@ public final class DiscoveryPostcardViewModel: DiscoveryPostcardViewModelType,
     self.userSessionEndedProperty.value = ()
   }
 
+  fileprivate let enableProjectCategoryExperiment = MutableProperty<Bool>(false)
```
Comment (Scollaco) [inline Library/ViewModels/DiscoveryPostcardViewModel.swift]: Another convention we use is to append the word `Property` to the MutableProperty name. This way, is easier to identify what's MutableProperty and what's signal on the code. So `enableProjectCategoryExperimentProperty` would be better.

File: Library/ViewModels/DiscoveryPostcardViewModelTests.swift [hunk 7]
```diff
@@ -483,4 +493,52 @@ internal final class DiscoveryPostcardViewModelTests: TestCase {
...
+    self.projectIsStaffPickViewHidden.assertValue(false)
+    self.projectCategoryStackViewHidden.assertValue(false)
+    self.projectCategoryName.assertValue(KsApi.Category.illustration.name)
+    self.projectCategoryViewHidden.assertValue(false)
+  }
+
+  func testHidesCategoryLabelExperimental() {
+    // Workaround for discrepancy between category ids from graphQL and category ids from the legacy API
+    let categoryId = KsApi.Category.illustration.intID
+    let illustrationCategory = KsApi.Category(id: String(categoryId!), name: KsApi.Category.illustration.name)
```
Comment (Scollaco) [inline Library/ViewModels/DiscoveryPostcardViewModelTests.swift]: To keep the code style consistent, this line can be replaced for its functional equivalent:
```
let illustrationCategory = KsApi.Category.illustration
                                |> KsApi.Category.lens.id .~ String(categoryId!) 
```

File: Kickstarter-iOS/Views/Cells/DiscoveryPostcardCell.swift [hunk 8]
```diff
@@ -44,12 +45,31 @@ internal final class DiscoveryPostcardCell: UITableViewCell, ValueCell {
...
+  fileprivate weak var projectIsStaffPickView: DiscoveryProjectCategoryView!
+
   private var projectSavedObserver: Any?
   private var sessionEndedObserver: Any?
   private var sessionStartedObserver: Any?
 
-    internal override func awakeFromNib() {
-    super.awakeFromNib()
+  internal override func awakeFromNib() {
+    if let categoryView = DiscoveryProjectCategoryView.fromNib(nib: Nib.DiscoveryProjectCategoryView) {
```
[the remark this answers is not in the document]
Comment (ifbarrera) [reply Kickstarter-iOS/Views/Cells/DiscoveryPostcardCell.swift]: `fromNib` can return nil so it's safer to use conditional binding.

File: Kickstarter-iOS/Views/Cells/DiscoveryPostcardCell.swift [hunk 9]
```diff
@@ -257,19 +291,27 @@ internal final class DiscoveryPostcardCell: UITableViewCell, ValueCell {
...
+
+    let showCategoriesExperiment =
+      Experiment.Name.showProjectCardCategory.isEnabled(in: AppEnvironment.current)
+
+    self.viewModel.inputs.enableProjectCategoryExperiment(showCategoriesExperiment)
   }
 
-  internal func configureWith(value: Project) {
-    self.viewModel.inputs.configureWith(project: value)
+  internal func configureWith(value: DiscoveryProjectCellRowValue) {
```
[the remark this answers is not in the document]
Comment (ifbarrera) [reply Kickstarter-iOS/Views/Cells/DiscoveryPostcardCell.swift]: Ah ok I didn't see that elsewhere so I wasn't sure if it was convention. Will do 👍

**Comment by Scollaco** [comment]
GIF

**Pull request merged**